org.eclipse.tm4e.languageconfiguration/src/main/java/org/eclipse/tm4e/languageconfiguration/ToggleLineCommentHandle.java

```diff
@@ -0,0 +1,245 @@
+			}
+			lineNumber++;
+		}
+		ITextSelection newSelection = new TextSelection(
+				selection.getOffset() - (isStartBeforeComment ? 0 : comment.length()),
+				selection.getLength() - deletedChars);
+		updateEditor(document, newTextbuilder.toString(), newSelection, editor);
+	}
+
+	private void addLineComments(IDocument document, ITextSelection selection, String comment, ITextEditor editor)
+			throws BadLocationException {
+		int lineNumber = selection.getStartLine();
+		int endLineNumber = selection.getEndLine();
+		StringBuilder newTextbuilder = new StringBuilder(document.get());
+		int insertedChars = 0;
+
+		while (lineNumber <= endLineNumber) {
+			newTextbuilder.insert(document.getLineOffset(lineNumber) + insertedChars, comment);
+			if (lineNumber != endLineNumber) {
+				insertedChars += comment.length();
+			}
+			lineNumber++;
+		}
+		ITextSelection newSelection = new TextSelection(selection.getOffset() + comment.length(),
+				selection.getLength() + insertedChars);
+		updateEditor(document, newTextbuilder.toString(), newSelection, editor);
+
+	}
+
+	private void removeBlockComment(IDocument document, ITextSelection selection, IRegion existingBlock,
+			CharacterPair blockComment, ITextEditor editor) {
+		int openOffset = existingBlock.getOffset();
+		int openLength = blockComment.getKey().length();
+		int closeOffset = existingBlock.getOffset() + existingBlock.getLength();
+		int closeLength = blockComment.getValue().length();
+		String text = document.get();
+		String newText = text.substring(0, openOffset) + text.substring(openOffset + openLength, closeOffset)
+				+ text.substring(closeOffset + closeLength);
+
+		int offsetFix = openLength;
+		int lengthFix = 0;
+		if (selection.getOffset() < openOffset + openLength) {
+			offsetFix = selection.getOffset() - openOffset;
+			lengthFix = openLength - offsetFix;
+		}
+		if (selection.getOffset() + selection.getLength() > closeOffset) {
+			lengthFix += selection.getOffset() + selection.getLength() - closeOffset;
+		}
+		ITextSelection newSelection = new TextSelection(selection.getOffset() - offsetFix,
+				selection.getLength() - lengthFix);
+		updateEditor(document, newText, newSelection, editor);
+	}
+
+	private void addBlockComment(IDocument document, ITextSelection selection, CharacterPair blockComment,
+			ITextEditor editor) {
+		String text = document.get();
+		String newText = text.substring(0, selection.getOffset()) + blockComment.getKey()
+				+ text.substring(selection.getOffset(), selection.getOffset() + selection.getLength())
+				+ blockComment.getValue() + text.substring(selection.getOffset() + selection.getLength());
+		ITextSelection newSelection = new TextSelection(selection.getOffset() + blockComment.getKey().length(),
+				selection.getLength());
+		updateEditor(document, newText, newSelection, editor);
+	}
+
+	private void updateEditor(IDocument document, String newText, ITextSelection newSelection, ITextEditor editor) {
+		document.set(newText);
```

Comment:
I am currently resetting the document to the new text. I forget if there is a proper way of doing this.

Reply:
You shhould see JDT code https://github.com/eclipse/eclipse.jdt.ui/blob/master/org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/actions/BlockCommentAction.java

It uses IRewriteTarget https://github.com/eclipse/eclipse.jdt.ui/blob/master/org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/actions/BlockCommentAction.java#L206 and IDocument#replace is called https://github.com/eclipse/eclipse.jdt.ui/blob/master/org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/actions/BlockCommentAction.java#L170